File: 23_merge_k_sorted_list/mergeKLists.py
class ListNode(object):
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution(object):
    def mergeTwoLists(self, l1, l2):
        result = res = ListNode(-1)
        while l1 and l2:
            if l2.val > l1.val:
                res.next = l1
                l1 = l1.next
            else:
                res.next = l2
                l2 = l2.next
            res = res.next

        res.next = l1 or l2
        return result.next
    
    def mergeKLists(self, lists):
        """
        :type lists: List[ListNode]
        """
        if not lists: return None
        
        first = 0
        last = len(lists) - 1
        middle = (first + last) / 2
        
        if len(lists) > 2:
            l1 = self.mergeKLists(lists[:middle+1])
            l2 = self.mergeKLists(lists[middle+1:])

        elif len(lists) == 2:
            l1 = lists[0]
            l2 = lists[1]

        else: return lists[0]

        res = self.mergeTwoLists(l1, l2)
        
        return res
if __name__ == '__main__':
    l1 = ListNode(1)
    l2 = ListNode(2)
    sol = Solution()
    node = sol.mergeTwoLists(l1, l2)
    while node:
        print node.val
        node = node.next

File: 23_merge_k_sorted_list/mergeKLists.cpp
#include <vector>
#include <iostream>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};


class Solution {
	public:
    	ListNode* mergeKLists(vector<ListNode*>& lists) {
            int first = 0, last = lists.size();
            if(!last) return NULL;
            mylists = lists;
            ListNode* res = mergeKListUtil(first, last);
            return res;
    	}
        ListNode* mergeKListUtil(int begin, int end) {
            int mid = (begin + end) / 2;
            ListNode* l1, *l2;
            if(end - begin > 2) {
                l1 = mergeKListUtil(begin, mid);
                l2 = mergeKListUtil(mid, end);
            }
            else if(end - begin == 2) {
                l1 = mylists[begin];
                l2 = mylists[begin + 1];
            }
            else return mylists[begin];

            return mergeTwoList(l1, l2);
        }
    private:
        ListNode* mergeTwoList(ListNode* l1, ListNode* l2) {
            ListNode* res = new ListNode(-1);
            ListNode* tmp = res;
            while(l1 && l2) {
                if(l2->val > l1->val) {
                    tmp -> next = l1;
                    l1 = l1 -> next;
                }
                else {
                    tmp -> next = l2;
                    l2 = l2 -> next;
                }
                tmp = tmp -> next;
            }
            tmp -> next = l1? l1 : l2;
            return res -> next;
        }
        vector<ListNode*> mylists;

};

int main() {
    int a = 5 / 2;
    cout << a << endl;
    /**
    int arr[] = {1, 2, 3, 4};
    vector<int> v(arr, arr + 4);
    vector<int> sub(&v[0], &v[2]);
    for(vector<int>::iterator it = sub.begin(); it != sub.end(); it++) {
        cout << *it << endl;
    }
    **/
}
